### PR TITLE
Migrate tests and coverage to GitHub Actions

### PR DESCRIPTION
**Pull request by crcrpar**
## Motivation

Per title.

## Description of the changes

- remove tests-python?? from .circleci/config.yml
- add coverage job to github actions
- update CONTRIBUTING.md as per the above changes

ref: #1928

### CODE REVIEW

**Review by HideakiImamura — APPROVED** [review]
Thanks for the PR as always. I confirmed that the `Tests/coverage` job is appropriately executed and `tests-python??` jobs are removed. LGTM.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -94,7 +94,11 @@ pytest tests/${TARGET_TEST_FILE_NAME}
 
 ## Continuous Integration and Local Verification
 
+Optuna repository uses GitHub Actions and CircleCI.
+
 CircleCI is used for continuous integration.
+Currently, we are migrating to GitHub Actions but still we use CirclCI for some tests listed below.
+`document` still exists because it makes it much easier to check built documentation and `tests-rdbstorage` is there because CircleCI better provides RDB storages.
```
Comment (hvy) [inline CONTRIBUTING.md:101]: How about explaining in what way CircleCI is "better", if not for the record here on GitHub?

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -71,3 +68,65 @@ jobs:
     - name: Tests
       run: |
         pytest tests --ignore tests/integration_tests
+
+  coverage:
```
Comment (hvy) [inline .github/workflows/tests.yml:72]: How about creating a separate workflow, e.g. `Tests (Coverage)` considering the fact that unit tests are already split depending on whether they're testing integration modules or not. Also, I think we can skip the cron scheduling for the code coverage.

**Review by hvy — COMMENTED** [review]
Thanks for the your PR as always. Left some comments!

File: CONTRIBUTING.md (hunk 1 above)
Comment (crcrpar) [reply CONTRIBUTING.md:101]: "CircleCI currently supports more various RDB storages than GitHub Actions" sounds better to you?

**Comment by crcrpar** [comment]
Okay, CircleCI documentation build job somehow failed, but I think it doesn't matter a lot.

**Comment by hvy** [comment]
Seemingly canceled somehow, let me rerun it.

File: .github/workflows/coverage.yml [hunk 3]
```diff
@@ -0,0 +1,70 @@
+name: Coverage
+
+on:
+  push:
+    - master
```
Comment (hvy) [inline .github/workflows/coverage.yml:5]: I think you need to specify the branches.
```suggestion
    branches:
      - master
```

**Comment by crcrpar** [comment]
Oh, somehow my push isn't reflected at all.

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -46,9 +46,6 @@ jobs:
         key: ${{ runner.os }}-${{ matrix.python-version }}-${{ env.cache-name }}-${{ hashFiles('**/setup.py') }}-v1
         restore-keys: |
           ${{ runner.os }}-${{ matrix.python-version }}-${{ env.cache-name }}-${{ hashFiles('**/setup.py') }}
-    - name: Setup environment
-      run: |
-        sudo apt-get update
```
Comment (hvy) [inline .github/workflows/tests.yml:51]: Is this change related to codecov?

File: .github/workflows/coverage.yml [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+      env:
+        OMP_NUM_THREADS: 1
+      run: |
+        pytest --cov=optuna --cov-report=xml tests
+
+    - name: Multi-node tests
+      env:
+        OMP_NUM_THREADS: 1
+      run: |
+        mpirun -n 2 -- pytest tests/integration_tests/test_chainermn.py
```
Comment (hvy) [inline .github/workflows/coverage.yml:66]: Probably outside the scope of this PR but I noticed that this pytest execution isn't recording the coverage?

File: .github/workflows/tests.yml (hunk 4 above)
Comment (crcrpar) [reply .github/workflows/tests.yml:51]: Negative. I thought this step should be removed as there's no apt install in this job. Am I stupid here?

File: .github/workflows/coverage.yml (hunk 5 above)
Comment (crcrpar) [reply .github/workflows/coverage.yml:66]: Okay, then remove this. I copied things from Circle CI config, sorry.

File: .github/workflows/tests.yml (hunk 4 above)
Comment (hvy) [reply .github/workflows/tests.yml:51]: We could separated it into its own PR if that's the case. But, it's a small change so it's not really a strong opinion if you'd be busy with other things. I by the way also think it's for the ChainerMN tests (integration tests which are not executed here).

**Review by hvy — APPROVED** [review]
LGTM!

Let me merge this PR since the changes look good although there are minor comments left.

File: .github/workflows/coverage.yml (hunk 5 above)
Comment (hvy) [reply .github/workflows/coverage.yml:66]: Created an issue at https://github.com/optuna/optuna/issues/2038.
